common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java

```diff
@@ -118,11 +119,18 @@ public ShuffleIndexInformation load(String filePath) throws IOException {
             return new ShuffleIndexInformation(filePath);
           }
         };
-    shuffleIndexCache = CacheBuilder.newBuilder()
-      .maximumWeight(JavaUtils.byteStringAsBytes(indexCacheSize))
-      .weigher((Weigher<String, ShuffleIndexInformation>)
-        (filePath, indexInfo) -> indexInfo.getRetainedMemorySize())
-      .build(indexCacheLoader);
+    CacheBuilder cacheBuilder = CacheBuilder.newBuilder()
+        .maximumWeight(JavaUtils.byteStringAsBytes(indexCacheSize))
+        .weigher((Weigher<String, ShuffleIndexInformation>)
+            (filePath, indexInfo) -> indexInfo.getRetainedMemorySize());
+    int expireTimeSeconds = conf.shuffleIndexCacheExpireTimeSeconds();
+    if (expireTimeSeconds > 0) {
+      shuffleIndexCache = cacheBuilder.expireAfterAccess(expireTimeSeconds, TimeUnit.SECONDS)
```

Comment:
Using the timeout threshold doesn't seem to be a friendly config for users. I don't think users could think of a reasonable timeout value to clean up the cache. In the worst case, they could result in regression if they set an arbitrary value, e.g., a very small value.

The only case that I think we can remove the shuffle cache now is when `applicationRemoved` is called. In that case, I think we could remove the shuffle cache safely, which could mitigate the issue you mentioned. However, I didn't see guava cache has such an API to remove unused cache entries.

Reply:
A small value of expireTimeSeconds is not what we want, and we can make sure it is greater than a fixed value (such as 300s)?